### PR TITLE
oauth2: add device flow support

### PR DESCRIPTION
**Pull request by marcosnils**
This is a rebased follow-up of https://github.com/golang/oauth2/pull/356. I'm opening a new PR since the original author doesn't seem to be active anymore on GH

Signed-off-by: Marcos Lilljedahl <marcosnils@gmail.com>

### CODE REVIEW

File: oauth2.go [hunk 1]
```diff
@@ -224,6 +226,63 @@ func (c *Config) Exchange(ctx context.Context, code string, opts ...AuthCodeOpti
...
+	return retrieveDeviceAuth(ctx, c, v)
+}
+
+// Poll does a polling to exchange an device code for a token.
+func (c *Config) Poll(ctx context.Context, da *DeviceAuth, opts ...AuthCodeOption) (*Token, error) {
+	v := url.Values{
+		"client_id":   {c.ClientID},
+		"grant_type":  {"urn:ietf:params:oauth:grant-type:device_code"},
+		"device_code": {da.DeviceCode},
+		"code":        {da.DeviceCode},
```
Comment (sealbro) [inline oauth2.go:250]: `code` - extra parameter can be removed
in this comment https://github.com/golang/oauth2/pull/356#discussion_r473582695

File: oauth2.go [hunk 2]
```diff
@@ -224,6 +226,63 @@ func (c *Config) Exchange(ctx context.Context, code string, opts ...AuthCodeOpti
...
+	}
+
+	// If no interval was provided, the client MUST use a reasonable default polling interval.
+	// See https://tools.ietf.org/html/draft-ietf-oauth-device-flow-07#section-3.5
+	interval := da.Interval
+	if interval == 0 {
+		interval = 5
+	}
+
+	for {
```
Comment (sealbro) [inline oauth2.go:266]: and need fix this comment https://github.com/golang/oauth2/pull/356#issuecomment-1295655154

Comment (sealbro) [reply oauth2.go:266]: I think the best solution would be to give the error or the token back, and the caller will decide how and how many times to call

File: oauth2.go [hunk 3]
```diff
@@ -224,6 +226,63 @@ func (c *Config) Exchange(ctx context.Context, code string, opts ...AuthCodeOpti
...
+
+		tok, err := retrieveToken(ctx, c, v)
+		if err == nil {
+			return tok, nil
+		}
+
+		errTyp := parseError(err)
+		switch errTyp {
+		case errAccessDenied, errExpiredToken:
+			return tok, errors.New("oauth2: " + errTyp)
```
Comment (sealbro) [inline oauth2.go:277]: It would be more convenient to immediately return a specialized error

File: oauth2.go (hunk 2 above)
Comment (marcosnils) [reply oauth2.go:266]: the spec doesn't say that an error should be returned if the interval is not particularly specified. In most of the cases, the auth service should return this value. Setting the default interval to a safe value seems to me that's ok.

File: oauth2.go (hunk 3 above)
Comment (marcosnils) [reply oauth2.go:277]: we are returning the error here. Not sure what the proposal currently is

**Comment by hickford** [comment]
Implemented in https://github.com/golang/oauth2/commit/e3fb0fb3af0ee7c0c62e31c393cbcfce6b2af5bf

**Pull request closed without merging**